### DISCUSSION TITLE
Hidden slides behaviour

### DISCUSSION THREAD

**Discussion by davidjerleke** [davidjerleke/embla-carousel#324] [Ideas]
OP by @MaxDesignFR:

> Hi @davidjerleke 
>
> About the documentation, it seems Embla carrousel does take into calculation hidden slides, which results in errors. Here is a very simple example I forked with 2 hidden slides: [https://codesandbox.io/s/embla-carousel-default-vanilla-forked-tsu2q2?file=/index.html](https://codesandbox.io/s/embla-carousel-default-vanilla-forked-tsu2q2?file=/index.html)
>
> Should I open a bug report for this, or maybe it could be mentioned in the docs? We can still manipulate the slides, but from what I've experienced, removing slides from the document flow such as `display:none` don't work, we actually need to remove them from the DOM entirely. 
>
> That's your call, the concise docs is actually really well made and easy to navigate, helped me give Embla a serious shot!

**Comment by davidjerleke** [comment]
Hi @MaxDesignFR,

Thank you for your input and a CodeSandbox that clearly demonstrates the issue. I'm curious, what's the use case for hidden slides, either by setting the `html hidden attribute` or using `display: none`?

Best,
David

**Reply by MaxDesignFR** [reply]
Hi @davidjerleke 

Appreciate your concern. So a simple use case would be a review carrousel with testimonies filtered by category (all reviews, client support reviews, reviews with images etc.). It offers a better UX when people use the slider since you can narrow down the slides, it's more practical and satisfying to use. So I think filtering slides whether it's text or images, may be the most obvious use case. But I could also imagine this being useful for efficient content A/B testing ; toggling the `display` property is quite easy. 

Here's an example of what I mean (I tried this with the current Embla v7) : https://i.vgy.me/sncJPG.gif 

To have it working, I clone a `template` tag with all slides, and when I want to filter it, I destroy the slider, remove all slides, rebuild it with desired slides from template, and reinstanciate Embla. I would rather not do all that workaround and toggle a class instead + call `reInit()` method, but it's doable though.

--- 

One remark for the v7. I find myself required to `justify-content: center; translate3d(0,0,0)!important` on the container (as well as hide prev/next buttons) when all slides together are less large than the viewport. Otherwise the caroussel does not center (as expected when there is not enough slides for the viewport size). I highlighted this quickly at the end of the above gif and made a quick demo: https://codesandbox.io/s/kbmzwd Could be nice to have an event when all slides together are less large than the viewport, so that we then can do our own custom css and js with it. I can open a new discussion for that if that helps making discussions organized, but on paper it looks like a request and I understand it's limited. 

Best regards

**Reply by davidjerleke** [reply]
Hi again @MaxDesignFR,

Thank you for the additional details.

> To have it working, I clone a template tag with all slides, and when I want to filter it, I destroy the slider, remove all slides, rebuild it with desired slides from template, and reinstanciate Embla. I would rather not do all that workaround and toggle a class instead + call reInit() method, but it's doable though.

It's not necessary to create a new instance. You can remove the slides that you want to be hidden from the DOM and run `embla.reInit()`. The same thing when you add slides to the DOM: calling `embla.reInit()` will be enough.

> it's more practical and satisfying to use

I agree that it would make it more convenient for Embla users. A follow up question then: How should Embla behave with the following API method:

- [`embla.slideNodes()`](https://www.embla-carousel.com/api/methods/#slidenodes)

Should it include hidden slides too or only visible slides? At the time of writing, I don't have an answer to this so feel free to give me your opinions. I think hidden slides should be excluded from these two methods though:

- [`embla.slidesInView()`](https://www.embla-carousel.com/api/methods/#slidesinview)
- [`embla.slidesNotInView()`](https://www.embla-carousel.com/api/methods/#slidesnotinview)

Best,
David

**Reply by MaxDesignFR** [reply]
> I agree that it would make it more convenient for Embla users. A follow up question then: How should Embla behave with the following API method:
> 
> [embla.slideNodes()](https://www.embla-carousel.com/api/methods/#slidenodes)

I think this would be preferable to include the hidden slides. Let's say I add a click event listener to the slides, I would want the hidden slides to listen to click event once they are displayed too. If I really need to filter those out, I could `embla.slideNodes().filter(slide => !slide.hasAttribute('hidden')` which is a very minor thing to do. 

I wonder if it should be an arbitrary decision for `embla.slidesInView()` and `embla.slidesNotInView()`, it really depends on what the user wants to do with the slides, including the hidden ones. But if for some reasons sticking to an arbitrary decision is best, I agree hidden slides should be excluded from those methods.

Otherwise I think a boolean parameter (whether or not to filter slides that are out of document flow) for each of these methods could cover all cases (I guess you'd have to determine if it's worth for the library as a whole, especially the lightweightness, I have no clue): 

Maybe something like:
• slideNodes : Parameters (hiddenSlides: boolean) - include hidden slides when `true`
• slidesInView : Parameters (target?: boolean, hiddenSlides: boolean) - include hidden slides when `true`
• slidesNotInView : Parameters (target?: boolean, hiddenSlides: boolean) - include hidden slides when `true`

Hope this feedback helps,
Max

---

> It's not necessary to create a new instance. You can remove the slides that you want to be hidden from the DOM and run embla.reInit(). The same thing when you add slides to the DOM: calling embla.reInit() will be enough.

Thanks for the tip, I actually forgot to add the options in reInit() and thought it wasn't working properly (loop option keeps working only if I add it again in the reInit() method).

**Reply by davidjerleke** [reply]
Hi @MaxDesignFR,

Thank you for your additional input. I’m not convinced that Embla should handle this. Because as you mentioned, people value the lightweightness quite high, and this is the first time someone brings the hidden slides matter up. Now this doesn’t mean that other users never thought about this before, but at least we can say that no one created an issue about this since this library was created. In contrast, many people have shared their appreciation about how lightweight this library is, and how extensible it is due to its rich API and plugin system.

Filtering hidden slides entirely from the API is quite easy, it’s actually just one line of code. But ignoring hidden slides from scroll snap calculations and including them in the API methods will increase the bundle size (as you suggest). And the upcoming v7 of Embla will already see an increase in bundle size because of the new convenient breakpoints option. Many users seem to accept an increased bundle size because of how much easier it will be to configure Embla with the breakpoints option. I’m not sure they feel the same about the hidden slides matter.

This is one of the main reasons why I added the plugin system to Embla: To enable users to extend Embla with features that aren’t built into the core.

Best,
David

**Reply by MaxDesignFR** [reply]
That makes complete sense. As I said I'm not aware if supporting that feature is heavy in term of code, so if it is I tend to agree with your reasoning, we better do this ourselves on top of the library. Thank you for considering this though.

**Reply by davidjerleke** [reply]
Hello @MaxDesignFR,

Just wanted to let you know that there’s a new option called [slides](https://www.embla-carousel.com/api/options/#slides) which enables you to filter slides.

Best,
David

**Reply by MaxDesignFR** [reply]
Thanks for the follow up! That's actually a very clever way to handle what we previously discussed, somehow we didn't think about this before (gonna try that soon). I suppose I'd need to reassign options with new slides css selector and then Embla.reInit(). Well done, really appreciate your work!

**Reply by davidjerleke** [reply]
Hi @MaxDesignFR,

1. For example, you can set the slides option to a [CSS not selector](https://developer.mozilla.org/en-US/docs/Web/CSS/:not) which filters out slides with the **hidden attribute**.

2. And I guess you can make it reactive with [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) by watching the hidden attribute on your slides, and call `reInit()` when needed.

I’m working on [this feature]( https://github.com/davidjerleke/embla-carousel/issues/159#issuecomment-1207342619) right now, and maybe that will result in step 2 being handled automatically, but at the time of writing, I’m not 100% sure how it will work.

Cheers,
David

**Reply by davidjerleke** [reply]
Hi @MaxDesignFR,

I finished the specifications yesterday. When I’m finished with the feature implementations I mentioned, you just have to follow [step 1 here](https://github.com/davidjerleke/embla-carousel/discussions/324#discussioncomment-5358211) and step 2 for you is to add the `embla-carousel-resize-observer` plugin which is being developed at the time of writing. Embla will automatically reInit as soon as you add/remove the hidden attribute then (because slide sizes change as soon as they are hidden or visible, which is why this will work).

Best,
David

**Reply by MaxDesignFR** [reply]
Hi,

A MutationObserver (for attribute changes) callback sounds perfect, in which I can update the `slides` option and then call `reInit()` method. So it sounds fairly simple to do, less than 10 lines of code. What else does `embla-carousel-resize-observer` do? Because if it do just that I feel like this could be in the inital Embla package (to support `slides` option change) but I may miss something here.

**Reply by davidjerleke** [reply]
Hi @MaxDesignFR,

> What else does `embla-carousel-resize-observer` do? Because if it do just that I feel like this could be in the inital Embla package (to support `slides` option change) but I may miss something here.

Nothing more actually. But most users value a small bundle size very much and many users don’t need this feature.

Good news though! I got a generous donation three days ago so I’ve been able to spend a decent amount of time refactoring parts of the Embla core. The result - The next version will automatically `reInit` when slide sizes (width/height) change and also when slides are added and removed, without a bundle size increase.

Stay tuned.

Best,
David
